simpleviewinc/keg-herkin
# Zen 403 jest test runner

**Ticket**: [issue_id](https://jira.simpleviewtools.com/browse/ZEN-403)

## Context

* We want to be able to run our end-to-end tests in a intractable environment
* Utilize playwright to run the same example tests in headless mode inside the docker container

## Goal
* Prove that we can execute the same tests in different test environments
* Configure jest to work inside `playwright` test setup
* Create example tests that showcases
   * Testing a dom element exists
   * Ensuring a value is set
   * Executing an action from a dom element
* The tests should be able to run in both the `keg-herkin` parent/child setup && without

### Update 1/19/21
```
* migrated the folder `/messenger/src/page` to `/messenger/src/parent/playwright/page`
* moved messenger.parent specific utils to `/messenger/src/parent/utils`
```

## Updates

* `container/Dockerfile` && `container/values.yml`
   * updated install syntax since we're not using alpine
   * utilizes `playwright`'s base image to set up our container so we can run playwright test inside the container

* `repos/backend/libs/fileSys/fileSys.js`
   * missing import && minor cleanup

* `repos/backend/libs/fileSys/testFiles.js`
   * `Limbo` result was returning an array. but we just care about the result string

* `repos/example/src/index.html`
   * added button that sets a text value

* `repos/messenger/src/parent/page/selectors.js`
   * updated to use the element wrapper for the selector methods

* `repos/messenger/src/utils/wrapper/elementWrapper.js`
  * wrapper for the `Playwright` API
  * could be expanded in the future if we end up needing the other methods

* `repos/tap/actions/init.js`
  * loads the example test data on init

* `repos/tap/actions/runner/upsertActiveData.js`
   * action that upserts the current active test content

* `repos/tap/app.js` && `repos/tap/containers/rootContainer.js`
   * moved the init action from `rootContainer` to `app.js`

* `repos/tap/components/aceEditor/aceEditor.js`
   * updated so it would change the content if `value` prop ever changes

* `repos/tap/containers/screens/runnerScreen.js`
  * updated to pull the content from the store instead of hardcoded

* `tests/tests/example/exampleTests.js`
  * moved the example test into its own file
  * added a new test that tests interaction and value changes

## Testing
**Setup**:
* NOTE: *I would recommend increasing the memory size of your docker desktop prior to pulling the image* if it's under 5 GB
    * [image](https://user-images.githubusercontent.com/3317835/104665733-5ae1ed80-568f-11eb-9141-dd381e53ff9f.png)

* Ensure you have latest `keg-cli` on `develop`

**Test 1:**
* Pull this branch locally
* run `yarn install`
* modify `/messenger/src/parent/parent.config.js` > `src` to http://herkin-zen-403-jest-test-runner.local.kegdev.xyz/
   * [image](https://user-images.githubusercontent.com/3317835/104665442-89ab9400-568e-11eb-97c5-af686a8323f9.png)

* run `keg herkin pack run package=docker.pkg.github.com/simpleviewinc/keg-packages/keg-herkin:zen-403-jest-test-runner port=5005:5005`
* run `yarn ex:mess` to start the example app with messenger
* Navigate to http://localhost:3000/
* expand the `keg-herkin` test editor 
* Verify you see the example test data
   * [image](https://user-images.githubusercontent.com/3317835/104665482-aa73e980-568e-11eb-85f5-a2663fd034a5.png)
* hit `run`
* Verify that the tests passes
    * [image](https://user-images.githubusercontent.com/3317835/104665518-cbd4d580-568e-11eb-8421-401dbfe3575a.png)


**Test 2:**
* attach to the container: `keg d att package-keg-herkin-zen-403-jest-test-runner`
* then run `yarn test:e2e`
    * this will run the tests within the `Playwright` context
    * it starts up the example app prior to executing the tests
* verify all the test passes on all 3 browsers
    * [image](https://user-images.githubusercontent.com/3317835/104665593-09d1f980-568f-11eb-84b1-8d79f32c0324.png)
